src/phylo/phylogeny.jl

```diff
@@ -0,0 +1,502 @@
+end
+
+@Docile.doc """
+Get one node by name.
+
+**Parameters:**
+* `tree`:  The Phylogeny to search.
+* `names`: The name of the nodes to get.
+""" ->
+function getindex(tree::Phylogeny, name::String)
+  for i in DepthFirst(tree)
+    if i.name == name
+      return i
+    end
+  end
+  error("No Node in phylogeny by specified name.")
+end
+
+@Docile.doc """
+Generate an index mapping names to nodes
+
+**Parameters:**
+* `tree`: The Phylogeny to index.
+""" ->
+function generateindex(tree::Phylogeny)
+  output = Dict{String, PhyNode}()
+  for i = BreadthFirst(tree)
+    if haskey(output, name(i))
+      error("You are trying to build an index dict " *
+            "of a tree with clades of the same name.")
+    end
+    output[name(i)] = i
+  end
+  return output
+end
+
+@Docile.doc """
+Find the shortest path between two nodes in a tree.
+
+**Parameters:**
+* `tree`: The Phylogeny to search in .
+* `n1`:   The first node.
+* `n2`:   The second node.
+""" ->
+function pathbetween(tree::Phylogeny, n1::PhyNode, n2::PhyNode)
+  if !isintree(tree, n1) || !isintree(tree, n2)
+    error("One of the nodes is not present in the tree.")
+  end
+  p1::Vector{PhyNode} = collect(Tip2Root(n1))
+  p2::Vector{PhyNode} = collect(Tip2Root(n2))
+  inter::Vector{PhyNode} = intersect(p1, p2)
+  filter!((x) -> !in(x, inter), p1)
+  filter!((x) -> !in(x, inter), p2)
+  return [p1, inter[1], reverse(p2)]
+end
+
+@Docile.doc """
+Find the distance between two nodes in a tree.
+
+**Parameters:**
+* `tree`: The Phylogeny to search in.
+* `n1`:   The first node.
+* `n2`:   The second node.
+""" ->
+function distance(tree::Phylogeny, n1::PhyNode, n2::PhyNode)
+  if !isintree(tree, n1) || !isintree(tree, n2)
+    error("One of the nodes is not present in the tree.")
+  end
+  p1::Vector{PhyNode} = collect(Tip2Root(n1))
+  p2::Vector{PhyNode} = collect(Tip2Root(n2))
+  inter::Vector{PhyNode} = intersect(p1, p2)
+  filter!((x) -> !in(x, inter), p1)
+  filter!((x) -> !in(x, inter), p2)
+  p = [p1, reverse(p2)]
+  return sum(distanceof, p)
+end
+
+@Docile.doc """
+Find the number of edges in the shortest path between two nodes in a tree.
+
+**Parameters:**
+* `tree`: The Phylogeny to search in.
+* `n1`: The first node.
+* `n2`: The second node.
+""" ->
+function depth(tree::Phylogeny, n1::PhyNode, n2::PhyNode)
+  p = pathbetween(tree, n1, n2)
+  return length(p) == 1 ? 0 : length(p) - 1
+end
+
+@Docile.doc """
+Find the distance between a node and the root of a tree.
+
+**Parameters:**
+* `tree`: The Phylogeny to search in.
+* `n1`:   The node.
+""" ->
+function distance(tree::Phylogeny, n1::PhyNode)
+  p = Tip2Root(n1)
+  return sum(getbranchlength, p)
+end
+
+@Docile.doc """
+Find the distance of each node from the root.
+
+**Parameters:**
+* `tree`: The Phylogeny to measure.
+""" ->
+function distance(tree::Phylogeny)
+  distances::TreeAnnotations{Float64} = TreeAnnotations{Float64}(tree)
+  function updatedistances(node::PhyNode, currentdist::Float64)
```

Comment:
This code can be inline. It doesn't seem like the function is helping anything, since it's only used once right after it's declared.